unittest/src/allow_db.rs

```diff
@@ -0,0 +1,216 @@
+    /// register values.
+    ///
+    /// The returned value is the tuple (address, len) passed into the
+    /// insert_rw_buffer call that created the RwAllowBuffer.
+    pub fn remove_rw_buffer(&mut self, buffer: RwAllowBuffer) -> (Register, Register) {
+        self.buffers.remove(&buffer.address);
+        (buffer.address.into(), buffer.len.into())
+    }
+}
+
+#[derive(Debug, PartialEq, thiserror::Error)]
+#[error("allow buffers overlap")]
+pub struct OverlapError;
+
+/// A read-only reference to a buffer that has been shared via the Allow system
+/// call. This reference is non-Copy, so `AllowDb` can determine when all
+/// references to the buffer have been destroyed.
+#[derive(Debug)]
+pub struct RoAllowBuffer {
+    // Safety invariant: Either length is 0, or address and length can be
+    // soundly converted to a &'static [u8].
+    address: *const u8,
+    len: usize,
+}
+
+// Allows access to the pointed-to-buffer. The returned reference has the same
+// lifetime as the &self reference, so the caller can't keep the reference for
+// longer than it has access to the RoAllowBuffer.
+impl std::ops::Deref for RoAllowBuffer {
+    type Target = [u8];
+    fn deref(&self) -> &[u8] {
+        match self.len {
+            0 => &[],
+            // Safety: Because length is nonzero, the safety invariant on
+            // address and len says this conversion is sound.
+            _ => unsafe { core::slice::from_raw_parts(self.address, self.len) },
+        }
+    }
+}
+
+/// A read-write reference to a buffer that has been shared via the Allow system
+/// call. This reference is non-Copy, so `AllowDb` can determine when all
+/// references to the buffer have been destroyed.
+#[derive(Debug)]
+pub struct RwAllowBuffer {
+    // Safety invariant: Either length is 0, or address and length can be safely
+    // converted to a &'static mut [u8].
+    address: *mut u8,
+    len: usize,
+}
+
+// Allows access to the pointed-to-buffer. The returned reference has the same
+// lifetime as the &self reference, so the caller can't keep the reference for
+// longer than it has access to the RwAllowBuffer.
+impl std::ops::Deref for RwAllowBuffer {
+    type Target = [u8];
+    fn deref(&self) -> &[u8] {
+        match self.len {
+            0 => &[],
+            // Safety: Because length is nonzero, the safety invariant on
+            // address and len says this conversion is sound.
+            _ => unsafe { core::slice::from_raw_parts(self.address, self.len) },
+        }
+    }
+}
+
+// Same purpose as the Deref implementation, but for mut references.
+impl std::ops::DerefMut for RwAllowBuffer {
+    fn deref_mut(&mut self) -> &mut [u8] {
+        match self.len {
+            0 => &mut [],
+            // Safety: Because length is nonzero, the safety invariant on
+            // address and len says this conversion is sound.
+            _ => unsafe { core::slice::from_raw_parts_mut(self.address, self.len) },
```

Comment:
Doesn't this also require that there not be another coexisting mutable reference to any memory overlapping this region? And if so, wouldn't that requirement also need to be placed on the address / len combo in RwAllowBuffer?

Reply:
Yes, that was intended to be part of the "can be soundly converted to a reference" part of the invariant on `RoAllowBuffer` and `RwAllowBuffer`. I added a sentence to each invariant (`RoAllowBuffer`'s and `RwAllowBuffer`'s) to clarify that aliasing restrictions are part of the invariant.

There's a subtle interaction here between TRD 104, `libtock_unittest`, and `libtock_platform`. TRD 104 does not prohibit processes from accessing buffers that have been shared with the kernel, although it allows userspace library implementations to do so. `libtock-rs` makes use of this prohibition, so that its Allow APIs can work with `&mut [u8]` and immutable references, rather than `&[Cell<u8>]` references.

I should have documented that explicitly. The safety requirement is actually on `RawSyscalls::syscall4`. Previously, the documentation said the safety requirements are inherited from the "underlying syscall" (meaning TRD 104), but this is a new requirement. I pushed a commit describing the new requirement.